p2p/protocol.go

```diff
@@ -304,19 +316,69 @@ func (c *conn) handleNewBlockHashes(ctx context.Context, msg ethp2p.Msg) error {
 
 	c.counter.WithLabelValues(packet.Name(), c.node.URLv4(), c.name).Add(float64(len(packet)))
 
-	hashes := make([]common.Hash, 0, len(packet))
-	for _, hash := range packet {
-		hashes = append(hashes, hash.Hash)
-		if err := c.getBlockData(hash.Hash); err != nil {
+	// Collect unique hashes for database write.
+	uniqueHashes := make([]common.Hash, 0, len(packet))
+
+	for _, entry := range packet {
+		hash := entry.Hash
+		if c.hasSeenBlockHash(hash) {
+			c.logger.Info().Str("hash", hash.Hex()).Msg("Skipping duplicate block hash")
+			continue
+		}
+
+		// **Attempt to fetch block data first**
+		if err := c.getBlockData(hash); err != nil {
 			return err
 		}
+
+		// **Now that we've successfully fetched, record the new block hash**
+		c.addBlockHash(hash)
+		uniqueHashes = append(uniqueHashes, hash)
 	}
 
-	c.db.WriteBlockHashes(ctx, c.node, hashes, tfs)
+	// Write only unique hashes to the database.
+	if len(uniqueHashes) > 0 {
+		c.db.WriteBlockHashes(ctx, c.node, uniqueHashes, tfs)
+	}
 
 	return nil
 }
 
+// addBlockHash adds a new block hash with a timestamp to the blockHashes list.
+func (c *conn) addBlockHash(hash common.Hash) {
+	now := time.Now()
+
+	// Remove all entries older than blockHashTTL
+	for e := c.blockHashes.Front(); e != nil; {
+		next := e.Next() // Save the next element
+		entry := e.Value.(BlockHashEntry)
+
+		if now.Sub(entry.time) <= blockHashTTL {
+			break
+		}
+
+		c.blockHashes.Remove(e)
+		e = next
+	}
```

Comment:
You should do this iteration in the same loop as in in `hasSeenBlockHash`. Could kill two birds one stone in that loop.

Reply:
good catch, added